xbmc/addons/interfaces/AudioEngine.cpp

```diff
@@ -0,0 +1,383 @@
+using namespace kodi; // addon-dev-kit namespace
+using namespace kodi::audioengine; // addon-dev-kit namespace
+
+namespace ADDON
+{
+
+void Interface_AudioEngine::Init(AddonGlobalInterface* addonInterface)
+{
+  addonInterface->toKodi->kodi_audioengine = (AddonToKodiFuncTable_kodi_audioengine*)malloc(sizeof(AddonToKodiFuncTable_kodi_audioengine));
+
+  // write KODI audio DSP specific add-on function addresses to callback table
+  addonInterface->toKodi->kodi_audioengine->make_stream = audioengine_make_stream;
+  addonInterface->toKodi->kodi_audioengine->free_stream = audioengine_free_stream;
+  addonInterface->toKodi->kodi_audioengine->get_current_sink_format = audioengine_get_current_sink_format;
+
+  // AEStream add-on function callback table
+  addonInterface->toKodi->kodi_audioengine->aestream_get_space = aestream_get_space;
+  addonInterface->toKodi->kodi_audioengine->aestream_add_data = aestream_add_data;
+  addonInterface->toKodi->kodi_audioengine->aestream_get_delay = aestream_get_delay;
+  addonInterface->toKodi->kodi_audioengine->aestream_is_buffering = aestream_is_buffering;
+  addonInterface->toKodi->kodi_audioengine->aestream_get_cache_time = aestream_get_cache_time;
+  addonInterface->toKodi->kodi_audioengine->aestream_get_cache_total = aestream_get_cache_total;
+  addonInterface->toKodi->kodi_audioengine->aestream_pause = aestream_pause;
+  addonInterface->toKodi->kodi_audioengine->aestream_resume = aestream_resume;
+  addonInterface->toKodi->kodi_audioengine->aestream_drain = aestream_drain;
+  addonInterface->toKodi->kodi_audioengine->aestream_is_draining = aestream_is_draining;
+  addonInterface->toKodi->kodi_audioengine->aestream_is_drained = aestream_is_drained;
+  addonInterface->toKodi->kodi_audioengine->aestream_flush = aestream_flush;
+  addonInterface->toKodi->kodi_audioengine->aestream_get_volume = aestream_get_volume;
+  addonInterface->toKodi->kodi_audioengine->aestream_set_volume = aestream_set_volume;
+  addonInterface->toKodi->kodi_audioengine->aestream_get_amplification = aestream_get_amplification;
+  addonInterface->toKodi->kodi_audioengine->aestream_set_amplification = aestream_set_amplification;
+  addonInterface->toKodi->kodi_audioengine->aestream_get_frame_size = aestream_get_frame_size;
+  addonInterface->toKodi->kodi_audioengine->aestream_get_channel_count = aestream_get_channel_count;
+  addonInterface->toKodi->kodi_audioengine->aestream_get_sample_rate = aestream_get_sample_rate;
+  addonInterface->toKodi->kodi_audioengine->aestream_get_data_format = aestream_get_data_format;
+  addonInterface->toKodi->kodi_audioengine->aestream_get_resample_ratio = aestream_get_resample_ratio;
+  addonInterface->toKodi->kodi_audioengine->aestream_set_resample_ratio = aestream_set_resample_ratio;
+}
+
+void Interface_AudioEngine::DeInit(AddonGlobalInterface* addonInterface)
+{
+  if (addonInterface->toKodi && /* <-- Safe check, needed so long old addon way is present */
+      addonInterface->toKodi->kodi_audioengine)
```

Comment:
The second check here is redundant.

Reply:
This has created here on Linux crashes, where on PVR and Game addons (still in old interfaces) and `addonInterface->toKodi` not becomes set (from here https://github.com/xbmc/xbmc/blob/master/xbmc/addons/binary-addons/AddonDll.cpp#L184).

On new becomes this used: https://github.com/xbmc/xbmc/blob/master/xbmc/addons/binary-addons/AddonDll.cpp#L264

Reply:
Maybe I understand wrong and you mean to leave `addonInterface->toKodi->kodi_audioengine` away? And let the check to `free` itself?

Reply:
@AlwinEsch yes, the latter. `free()` (or `delete`) will do nothing on a null pointer. `clang-tidy` [complains](https://clang.llvm.org/extra/clang-tidy/checks/readability-delete-null-pointer.html) if we check this. :smile:

Reply:
Have removed the `addonInterface->toKodi->kodi_audioengine` check